src/object_store/cache.rs

```diff
@@ -217,88 +223,201 @@ impl CachingObjectStore {
+                debug!("{path} {chunk_range:?} starting");
+                let mut batch_data = match self.inner.get_range(path, batch_range).await {
+                    Ok(data) => data,
+                    Err(err) => {
+                        debug!("{path} {chunk_range:?} error: {err}");
+                        // object store error is not `Clone`, so we need to do this
+                        let message = format!("{err}");
+                        // Propagate error to all awaiting tasks.
+                        for (_, tx) in chunk_batches {
+                            let _ = tx.send(Err(super::ObjectStoreError::Generic {
+                                reason: message.clone(),
+                            }));
+                        }
+                        return Err(object_store::Error::Generic {
+                            store: "cache_store",
+                            source: Box::new(super::ObjectStoreError::Generic {
+                                reason: message,
+                            }),
+                        });
+                    }
+                };
+                debug!("{path} {chunk_range:?} completed");
+
+                // `Bytes` are `cheaply cloneable and sliceable chunk of contiguous memory`, so
+                // there should be no need to wrap them in an `Arc`.
+                result.put(batch_data.clone());
+
+                // TODO: execute this in a separate task
+                for (key, tx) in chunk_batches {
+                    // Split the next chunk from the batch
+                    let data = if batch_data.len() < self.min_fetch_size as usize {
+                        batch_data.clone()
+                    } else {
+                        batch_data.split_to(self.min_fetch_size as usize)
+                    };
+
+                    // Cache the memory value
+                    self.cache
+                        .insert(key.clone(), CacheValue::Memory(data.clone()))
+                        .await;
+                    // Send out to any threads awaiting on this entry
+                    if let Err(err) = tx.send(Ok(data.clone())) {
+                        // This is ok, it just means there weren't any concurrent threads trying to
+                        // access this key in-between this thread putting the `CacheValue::Pending`
+                        // in, and evaluating it and replacing it with the `CacheValue::Memory`
+                        debug!("Failed to send chunk batch data for {key:?}: {err}");
+                    };
+
+                    // Finally trigger persisting to disk
+                    let cache = self.cache.clone();
+                    let file_manager = self.file_manager.clone();
+                    tokio::task::spawn(async move {
+                        // Run pending tasks to avert eviction races.
+                        cache.run_pending_tasks().await;
```

Comment:
Should there be some kind of a critical section/lock? Intuitively, we're in a concurrent context and so whatever state we put the cache in after this `run_pending_tasks` might not stay the same when we reach the next line. What does the eviction race look like here?

Reply:
> What does the eviction race look like here?

It looks like:
- moka starting the eviction future
- a new entry comes in and gets persisted to disk
- the eviction future completes and deletes the above new entry (but the file pointer in the cache stays)
- the cached file read errors out

There are some nuances above (different outcomes depending on whether we error out on writing the cache file or not).